contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+      voteToken,
+      candidateFee,
+      active: false,
+      creator: api.sender,
+      lastTickTime: now.getTime(),
+    };
+    const insertedInst = await api.db.insert('instances', newInstance);
+
+    if (api.sender !== api.owner
+        && api.sender !== 'null'
+        && api.BigNumber(instanceCreationFee).gt(0)) {
+      await api.executeSmartContract('tokens', 'transfer', {
+        // eslint-disable-next-line no-template-curly-in-string
+        to: 'null', symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'", quantity: instanceCreationFee, isSignedWithActiveKey,
+      });
+    }
+    api.emit('createInstance', { id: insertedInst._id });
+  }
+};
+
+actions.updateInstance = async (payload) => {
+  const {
+    instanceId, candidateFee, isSignedWithActiveKey,
+  } = payload;
+
+  const params = await api.db.findOne('params', {});
+  const { instanceUpdateFee } = params;
+
+  // eslint-disable-next-line no-template-curly-in-string
+  const utilityTokenBalance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'" });
+
+  const authorizedUpdate = api.BigNumber(instanceUpdateFee).lte(0) || api.sender === api.owner
+    ? true
+    : utilityTokenBalance && api.BigNumber(utilityTokenBalance.balance).gte(instanceUpdateFee);
+
+  if (api.assert(authorizedUpdate, 'you must have enough tokens to cover the update fee')
+    && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    && api.assert(typeof instanceId === 'string' && api.BigNumber(instanceId).isInteger(), 'invalid instanceId')
+    && api.assert(candidateFee, 'specify at least one field to update')) {
+    const existingInst = await api.db.findOne('instances', { _id: api.BigNumber(instanceId).toNumber() });
+    if (!api.assert(existingInst, 'instance not found')
+      || !api.assert(existingInst.creator === api.sender || api.owner === api.sender, 'must be instance creator')) return;
+
+    if (candidateFee) {
+      if (!api.assert(typeof candidateFee === 'object'
+        && typeof candidateFee.method === 'string'
+        && FeeMethod.indexOf(candidateFee.method) !== -1
+        && typeof candidateFee.symbol === 'string'
+        && typeof candidateFee.amount === 'string' && api.BigNumber(candidateFee.amount).gte(0), 'invalid candidateFee object')) return;
+      const feeTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: candidateFee.symbol });
+      if (!api.assert(feeTokenObj && api.BigNumber(candidateFee.amount).dp() <= feeTokenObj.precision, 'invalid candidateFee token or precision')) return;
+      existingInst.candidateFee = candidateFee;
+    }
+
+    await api.db.update('instances', existingInst);
+
+    // burn the token update fees
+    if (api.sender !== api.owner
+        && api.sender !== 'null'
+        && api.BigNumber(instanceUpdateFee).gt(0)) {
+      await api.executeSmartContract('tokens', 'transfer', {
```

Comment:
Same as above, should handle the fees and verify success before updating the instance.